packages/php/email-editor/src/Engine/Renderer/class-html2text.php

```diff
@@ -0,0 +1,622 @@
+		if ( ! is_string( $html ) ) {
+			$html = (string) $html;
+		}
+
+		$doc = self::get_document( $html, $options );
+
+		$output = self::iterate_over_node( $doc, null, false, $is_office_document, $options );
+
+		// Process output for whitespace/newlines.
+		$output = self::process_whitespace_newlines( $output );
+
+		return $output;
+	}
+
+	/**
+	 * Unify newlines
+	 *
+	 * Converts \r\n to \n, and \r to \n. This means that all newlines
+	 * (Unix, Windows, Mac) all become \ns.
+	 *
+	 * @param string $text Text with any number of \r, \r\n and \n combinations.
+	 * @return string The fixed text.
+	 */
+	public static function fix_newlines( string $text ): string {
+		// Replace \r\n to \n.
+		$text = str_replace( "\r\n", "\n", $text );
+		// Remove \rs.
+		$text = str_replace( "\r", "\n", $text );
+
+		return $text;
+	}
+
+	/**
+	 * Get non-breaking space character codes
+	 *
+	 * @return array<string> Array of nbsp codes.
+	 */
+	public static function nbsp_codes(): array {
+		return array(
+			"\xc2\xa0",
+			"\u00a0",
+		);
+	}
+
+	/**
+	 * Get zero-width non-joiner character codes
+	 *
+	 * @return array<string> Array of zwnj codes.
+	 */
+	public static function zwnj_codes(): array {
+		return array(
+			"\xe2\x80\x8c",
+			"\u200c",
+		);
+	}
+
+	/**
+	 * Remove leading or trailing spaces and excess empty lines from provided multiline text
+	 *
+	 * @param string $text Multiline text with any number of leading or trailing spaces or excess lines.
+	 * @return string The fixed text.
+	 */
+	public static function process_whitespace_newlines( string $text ): string {
+
+		// Remove excess spaces around tabs.
+		$result = preg_replace( '/ *\t */im', "\t", $text );
+		$text   = null !== $result ? $result : $text;
+
+		// Remove leading whitespace.
+		$text = ltrim( $text );
+
+		// Remove leading spaces on each line.
+		$result = preg_replace( "/\n[ \t]*/im", "\n", $text );
+		$text   = null !== $result ? $result : $text;
+
+		// Convert non-breaking spaces to regular spaces to prevent output issues,
+		// do it here so they do NOT get removed with other leading spaces, as they
+		// are sometimes used for indentation.
+		$text = self::render_text( $text );
+
+		// Remove trailing whitespace.
+		$text = rtrim( $text );
+
+		// Remove trailing spaces on each line.
+		$result = preg_replace( "/[ \t]*\n/im", "\n", $text );
+		$text   = null !== $result ? $result : $text;
+
+		// Unarmor pre blocks.
+		$text = self::fix_newlines( $text );
+
+		// Remove unnecessary empty lines.
+		$result = preg_replace( "/\n\n\n*/im", "\n\n", $text );
+
+		return null !== $result ? $result : $text;
+	}
+
+	/**
+	 * Can we guess that this HTML is generated by Microsoft Office?
+	 *
+	 * @param string $html The HTML content.
+	 * @return bool True if this appears to be an Office document.
+	 */
+	public static function is_office_document( string $html ): bool {
+		return strpos( $html, 'urn:schemas-microsoft-com:office' ) !== false;
```

Comment:
Using `strpos` instead of `str_contains` here (and `str_starts_with` elsewhere in the file) because of the PHP 7.4 compatability requirement.